src/cls/journal/cls_journal.cc

```diff
@@ -224,6 +224,86 @@ int expire_tags(cls_method_context_t hctx, const std::string *skip_client_id) {
   return 0;
 }
 
+int get_client_list_range(cls_method_context_t hctx,
+			  std::set<cls::journal::Client> &clients,
+			  std::string start_after, uint64_t max_return) {
+  std::string last_read;
+  if (!start_after.empty()) {
+    last_read = key_from_client_id(start_after);
+  }
+
+  std::map<std::string, bufferlist> vals;
+  int r = cls_cxx_map_get_vals(hctx, last_read, HEADER_KEY_CLIENT_PREFIX,
+                               max_return, &vals);
+  if (r < 0) {
+    CLS_ERR("failed to retrieve omap values: %s", cpp_strerror(r).c_str());
+    return r;
+  }
+
+  for (std::map<std::string, bufferlist>::iterator it = vals.begin();
+       it != vals.end(); ++it) {
+    try {
+      bufferlist::iterator iter = it->second.begin();
+
+      cls::journal::Client client;
+      ::decode(client, iter);
+      clients.insert(client);
+    } catch (const buffer::error &err) {
+      CLS_ERR("could not decode client '%s': %s", it->first.c_str(),
+              err.what());
+      return -EIO;
+    }
+  }
+
+  return 0;
+}
+
+int find_min_commit_position(cls_method_context_t hctx,
+			     cls::journal::ObjectSetPosition &minset) {
+  int r;
+  std::string start_after;
+  std::set<cls::journal::Client> clients;
+  std::set<cls::journal::Client> batch;
+
+  start_after = "";
+  do {
+    batch.clear();
+    r = get_client_list_range(hctx, batch, start_after, JOURNAL_MAX_RETURN);
+    if (r == 0) {
+      if (!batch.empty()) {
+	start_after = batch.rbegin()->id;
+	clients.insert(batch.begin(), batch.end());
+      }
+    }
+  } while((r == 0) && !(batch.size() < JOURNAL_MAX_RETURN));
+
+  if (r < 0) {
+    return r;
+  }
+
+  bool valid = false;
+  uint64_t tag_tid = 0, entry_tid = 0;
+  cls::journal::ObjectPosition first;
+  cls::journal::ObjectSetPosition object_set_position;
+
+  for (std::set<cls::journal::Client>::iterator it = clients.begin();
```

Comment:
Move this section up into the ```do { } while``` batch loop so that you have a restricted working set.